Sources/NIOCore/Docs.docc/swift-concurrency.md

```diff
@@ -0,0 +1,244 @@
+new types. Importantly, these types need to uphold the channel's back-pressure and writability guarantees.
+NIO introduced the ``NIOThrowingAsyncSequenceProducer``, ``NIOAsyncSequenceProducer`` and the ``NIOAsyncWriter``
+which form the foundation to bridge a ``Channel``.
+On top of these foundational types, NIO provides the `NIOAsyncChannel` which is used to wrap a 
+``Channel`` to produce an interface that can be consumed directly from Swift Concurrency. The following
+sections cover the details of the foundational types and how the `NIOAsyncChannel` works.
+
+#### NIOThrowingAsyncSequenceProducer and NIOAsyncSequenceProducer
+
+The ``NIOThrowingAsyncSequenceProducer`` and ``NIOAsyncSequenceProducer`` are root asynchronous sequences
+similar to Swift's `AsyncStream`. Their core goal is to provide a back-pressured bridge between a
+synchronous producer and an asynchronous consumer. These types are highly configurable and generic which
+makes them usable in a lot of places with very good performance; however, at the same time they are
+not the easiest types to hold. We recommend that you **never** expose them in public API but rather
+wrap them in your own async sequence.
+
+#### NIOAsyncWriter
+
+The ``NIOAsyncWriter`` is used for bridging from an asynchronous producer to a synchronous consumer.
+It also has back-pressure support which allows the synchronous side to suspend calls to the
+``NIOAsyncWriter/yield(contentsOf:)`` method.
+
+
+> Important: Everything below this is currently not public API but can be test it by using `@_spi(AsyncChannel) import`.
+The APIs might change until they become publicly available.
+
+#### NIOAsyncChannel
+
+The above types are used to bridge both the read and write side of a ``Channel`` into Swift Concurrency.
+This can be done by wrapping a ``Channel`` via the `NIOAsyncChannel/init(synchronouslyWrapping:backpressureStrategy:isOutboundHalfClosureEnabled:inboundType:outboundType:)`
+initializer. Under the hood, this initializer adds two channel handlers to the end of the channel's pipeline.
+These handlers bridge the read and write side of the channel. Additionally, the handlers are working together
+to close the channel once both the reading and the writing have finished.
+
+
+This is how you can wrap an existing channel into a `NIOAsyncChannel`, consume the inbound data and
+echo it back outbound.
+
+```swift
+let channel = StringChannel()
+let asyncChannel = try NIOAsyncChannel(synchronouslyWrapping: channel, inboundType: String.self, outboundType: String.self)
+
+for try await inboundData in asyncChannel.inboundStream {
+    try await asyncChannel.outboundWriter.write(inboundData)
+}
+```
+
+### Async bootstrap
+
+The `NIOAsyncChannel` alone would already allow you to wrap your channels and use them from Concurrency;
+however, you must be very careful at what point you wrap your channel otherwise you might lose some reads.
```

Comment:
I think you need to elaborate on this point if you're going to bring it up here. Why might you lose reads? Under what circumstances?

Reply:
Presumably using an async bootstrap helps you to avoid this so having this at the end of the previous section would sets up the reader nicely for this section nicely.